pkg/service/service.go

```diff
@@ -431,6 +431,50 @@ func (s *service) WatchModel(ctx context.Context, ns resource.Namespace, modelID
 	return state, message, nil
 }
 
+// checkRequesterPermission validates that the authenticated user can make
+// requests on behalf of the resource identified by the requester UID.
+func (s *service) checkRequesterPermission(ctx context.Context, model *datamodel.Model) error {
+	authType := resource.GetRequestSingleHeader(ctx, constant.HeaderAuthTypeKey)
+	if authType != "user" {
+		// Only authenticated users can switch namespaces.
+		return ErrUnauthenticated
+	}
+
+	requester := resource.GetRequestSingleHeader(ctx, constant.HeaderRequesterUIDKey)
+	authenticatedUser := resource.GetRequestSingleHeader(ctx, constant.HeaderUserUIDKey)
+	if requester == "" || authenticatedUser == requester {
+		// Request doesn't contain impersonation.
+		return nil
+	}
+
+	// The only impersonation that's currently implemented is switching to an
+	// organization namespace.
+	isMember, err := s.aclClient.CheckPermission(ctx, "organization", uuid.FromStringOrNil(requester), "member")
+	if err != nil {
+		return errmsg.AddMessage(
+			fmt.Errorf("checking organization membership: %w", err),
```

Comment:
ℹ️ We should consider adding an `errmsg` inteceptor in the middleware to return the end-user messages in the API & console (not in the scope of this PR). We were discussing how to improve error messages across the system yesterday with product.

Reply:
Totally, if you can tag me in the the discussion, I can arrange my time to have it in model-backend